controllers/capabilities/tenant_controller.go

```diff
@@ -95,119 +107,116 @@ func (r *TenantReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
-				if err != nil {
-					r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "Failed to delete tenant", "%v", err)
-					return ctrl.Result{}, err
-				}
-			} else {
-				reqLogger.Info("Removing tenant CR - tenant is already scheduled for deletion", "tenantID", existingTenant.Signup.Account.ID)
-			}
+		// On wait error, retry
+		if helper.IsWaitError(reconcileErr) {
+			reqLogger.Info("ERROR", "wait error", reconcileErr)
+			return ctrl.Result{Requeue: true}, nil
 		}
 
-		controllerutil.RemoveFinalizer(tenantR, tenantFinalizer)
-		err = r.UpdateResource(tenantR)
-		if err != nil {
-			return ctrl.Result{}, err
+		// On orphan error, do not retry
+		if helper.IsOrphanSpecError(reconcileErr) {
+			reqLogger.Info("ERROR", "orphan error", reconcileErr)
+			return ctrl.Result{Requeue: false}, nil
 		}
 
-		return ctrl.Result{}, nil
+		// On all other errors, retry
+		reqLogger.Error(reconcileErr, "failed to reconcile")
+		r.EventRecorder().Eventf(tenantR, corev1.EventTypeWarning, "ReconcileError", "%v", reconcileErr)
+		return ctrl.Result{}, reconcileErr
 	}
 
-	// Ignore deleted resources, this can happen when foregroundDeletion is enabled
-	// https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion
-	if tenantR.GetDeletionTimestamp() != nil {
-		return ctrl.Result{}, nil
-	}
+	reqLogger.Info("Tenant reconciled successfully")
+	return ctrl.Result{}, nil
+}
 
-	// If the tenant.Status.TenantID is found and the annotation is not found - create
-	// If the tenant.Status.TenantID is found and the annotation is found but, the value of annotation is different to the status.TenantID - update
-	tenantIdAnnotationFound := true
-	tenantId := tenantR.Status.TenantId
-	if value, found := tenantR.ObjectMeta.Annotations[tenantIdAnnotation]; tenantId != 0 && !found || tenantId != 0 && found && value != strconv.FormatInt(tenantR.Status.TenantId, 10) {
-		if tenantR.ObjectMeta.Annotations == nil {
-			tenantR.ObjectMeta.Annotations = make(map[string]string)
-		}
+func (r *TenantReconciler) specReconciler(tenantCR *capabilitiesv1alpha1.Tenant, logger logr.Logger) (*TenantStatusReconciler, error) {
+	// Save a copy of the original tenant CR prior to changing it for status reconciler
+	originalTenantCR := tenantCR.DeepCopy()
 
-		tenantR.ObjectMeta.Annotations[tenantIdAnnotation] = strconv.FormatInt(tenantR.Status.TenantId, 10)
-		tenantIdAnnotationFound = false
+	// Setup porta client
+	portaClient, err := r.setupPortaClient(tenantCR, originalTenantCR, logger)
+	if err != nil {
+		statusReconciler := NewTenantStatusReconciler(r.BaseReconciler, tenantCR, originalTenantCR, err)
+		return statusReconciler, err
 	}
 
-	tenantFinalizerFound := true
-	if !controllerutil.ContainsFinalizer(tenantR, tenantFinalizer) {
-		controllerutil.AddFinalizer(tenantR, tenantFinalizer)
-		tenantFinalizerFound = false
+	// Process tenant deletion
+	tenantInDeletion, err := r.reconcileDeletion(tenantCR, portaClient, logger)
```

Comment:
reconcile deletion in the `SpecReconciler` ??? I think this should be done elsewhere.

Reply:
I agree. But currently I'm doing it in mind with adding unit tests. Ideally the delete should be one of the first things to do in the main reconciler however, since the tenant controller requires porta for deletion, we would need to first progress the epic around mocking porta on top level reconciler (since we already got a way of doing it within functions called from the reconciler I moved it here to be able to mock it with what we have).

So my only alternative is to move it to the main reconciler (as it was) without unit testing it until we get the tests setup in place.

Reply:
Updated, thank you